p2p/peerexchange/bookchecker_test.go

```diff
@@ -154,6 +155,93 @@ func TestDiscovery_GetRandomPeers(t *testing.T) {
 	})
 }
 
+func TestHost_ReceiveAddressesOnCheck(t *testing.T) {
+	t.Parallel()
+
+	hosts, instances := setupOverNodes(t, 2)
+	nodeA, nodeB := instances[0], instances[1]
+
+	best, err := bestHostAddress(nodeA.host)
+	require.NoError(t, err)
+
+	connectedNodeAddress := addressFromHost(t, hosts[1])
+	nodeA.book.AddAddress(connectedNodeAddress, best)
+	wrapDiscovery(instances)
+
+	// check that after bootstrap, nodeB has only addresss of nodeA and nodeA has only addresss of nodeB
+	require.Eventually(t, func() bool {
+		aAddresses := nodeA.book.GetAddresses()
+		bAddresses := nodeB.book.GetAddresses()
+		if len(aAddresses) != 1 || len(bAddresses) != 1 {
+			return false
+		}
+		return aAddresses[0].ID.String() == nodeB.host.ID().String() && bAddresses[0].ID.String() == nodeA.host.ID().String()
+	}, 4*time.Second, 100*time.Millisecond, "nodeA should have only addresss of nodeB and nodeB should have only addresss of nodeA")
+
+	// disconnect nodeA and node. sleep 3 sec need for addresses become outdated and ready for check
+	require.NoError(t, nodeB.host.Network().ClosePeer(nodeA.host.ID()))
+	time.Sleep(3 * time.Second)
+
+	// add new nodes to network
+	hostsAdditional, _ := setupOverNodes(t, 2)
+	nodeA.book.AddAddress(addressFromHost(t, hostsAdditional[0]), best)
+	nodeA.book.AddAddress(addressFromHost(t, hostsAdditional[1]), best)
+	require.Eventually(t, func() bool {
+		addresses := nodeA.book.AddressCache()
+		if len(addresses) != 3 {
+			return false
+		}
+		for _, address := range addresses {
+			valid := address.ID.String() == hostsAdditional[0].ID().String() ||
+				address.ID.String() == hostsAdditional[1].ID().String() ||
+				address.ID.String() == hosts[1].ID().String()
+			if !valid {
+				return false
+			}
```

Comment:
same here. once we are sure about the size. we should exit require.Eventually() and check the expected values outside.